File: python/interviewAlgorithms/isCycle.py
# class Node:
#     def __init__(self,data):
#         self.data = data
#         self.next = None

# class IsCycle:
#     @staticmethod
#     def isCycle(head):
#         p1 = head
#         p2 = head
#         while p2 is not None and p2.next is not None:
#             p1 = p1.next
#             p2 = p2.next.next
#             if p1 == p2:
#                 return True
#         return False
    
# if __name__ =="__main__":
#     node1 = Node(5)
#     node2 = Node(3)
#     node3 = Node(7)
#     node4 = Node(2)
#     node5 = Node(6)
#     node6 = Node(8)
#     node7 = Node(1)
#     node1.next = node2
#     node2.next = node3
#     node3.next = node4
#     node4.next = node5
#     node5.next = node6
#     node6.next = node7
#     node7.next = node4
#     print(IsCycle.isCycle(node1))
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
    
    def append(self,data):
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = new_node
    
    def append_node(self,node):
        if self.head is None:
            self.head = node
        else:
            current = self.head
            while current.next:
                current = current.next
            current.next = node

            #将末尾节点的next指针指向头节点，形成环
            node.next = self.head

    def is_cycle(self):
        p1 = self.head
        p2 = self.head
        cycle_found =False

        while p2 and p2.next:
            p1 = p1.next
            p2 = p2.next.next

            if p1 == p2:
                cycle_found =True
                break
        
        if not cycle_found:
            return False

        #计算环的长度
        length = 1
        p1 = p1.next
        while p1 != p2:
            p1 = p1.next
            length +=1
        
        # 找到入环点
        p1 = self.head
        while p1 != p2:
            p1 = p1.next
            p2 = p2.next

        return length, p1.data


if __name__ == "__main__":
    linked_list = LinkedList()
    linked_list.append(5)
    linked_list.append(3)
    linked_list.append(7)
    linked_list.append(2)
    linked_list.append(6)
    linked_list.append(8)
    linked_list.append(1)

    #创建环
    node4 = linked_list.head.next.next.next
    linked_list.append_node(node4)

    cycle_length,entry_point = linked_list.is_cycle()
    print("Cycle Length:",cycle_length)
    print("Entry Point:",entry_point)

File: C++/interviewAlgorithms/isCycle.cpp
#include<iostream>
using namespace std;

class Node {
public:
    int data;
    Node* next;
    Node(int data) {
        this->data = data;
        this->next = nullptr;
    }
};

class isCycle {
public:
    static Node* findMeetingPoint(Node* head) {
        Node* slow = head;
        Node* fast = head;
        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast) {
                return slow;  // 返回相遇点
            }
        }
        return nullptr;  // 不存在环，返回nullptr
    }

    static Node* findEntrypoint(Node* head) {
        Node* meetingPoint = findMeetingPoint(head);

        if (meetingPoint == nullptr) {  // 不存在环，返回nullptr
            return nullptr;
        }

        Node* p1 = head;
        Node* p2 = meetingPoint;

        while (p1 != p2) {
            p1 = p1->next;
            p2 = p2->next;
        }
        return p1;  // 返回入环点
    }

    static int calculateCycleLength(Node* head) {
        Node* meetingPoint = findMeetingPoint(head);

        if (meetingPoint == nullptr) {  // 不存在环，返回0
            return 0;
        }

        Node* p1 = meetingPoint;
        int length = 1;

        while (p1->next != meetingPoint) {
            p1 = p1->next;
            length++;
        }

        return length;  // 返回环的长度
    }
};

int main() {
    Node* node1 = new Node(5);
    Node* node2 = new Node(3);
    Node* node3 = new Node(7);
    Node* node4 = new Node(2);
    Node* node5 = new Node(6);
    Node* node6 = new Node(8);
    Node* node7 = new Node(1);
    node1->next = node2;
    node2->next = node3;
    node3->next = node4;
    node4->next = node5;
    node5->next = node6;
    node6->next = node7;
    node7->next = node4;

    Node* entryPoint = isCycle::findEntrypoint(node1);
    int cycleLength = isCycle::calculateCycleLength(node1);

    if (entryPoint != nullptr) {
        cout << "入环点为： " << entryPoint->data << endl;
        cout << "环的长度为： " << cycleLength << endl;
    }
    else {
        cout << "链表中不存在环" << endl;
    }
    return 0;
}
